controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,342 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/controller/utils"
+	"github.com/flynn/flynn/host/types"
+)
+
+const eventBufferSize int = 1000
+
+type Scheduler struct {
+	utils.ControllerClient
+	utils.ClusterClient
+	log        log15.Logger
+	formations *Formations
+
+	jobs map[string]*Job
+
+	listeners map[chan Event]struct{}
+	listenMtx sync.RWMutex
+
+	stop     chan struct{}
+	stopOnce sync.Once
+
+	formationChange chan *ct.ExpandedFormation
+	jobRequests     chan *JobRequest
+
+	validJobStatuses map[host.JobStatus]bool
+}
+
+func NewScheduler(cluster utils.ClusterClient, cc utils.ControllerClient) *Scheduler {
+	return &Scheduler{
+		ControllerClient: cc,
+		ClusterClient:    cluster,
+		log:              log15.New("component", "scheduler"),
+		jobs:             make(map[string]*Job),
+		listeners:        make(map[chan Event]struct{}),
+		stop:             make(chan struct{}),
+		formations:       newFormations(),
+		formationChange:  make(chan *ct.ExpandedFormation, 1),
+		jobRequests:      make(chan *JobRequest, eventBufferSize),
+		validJobStatuses: map[host.JobStatus]bool{
+			host.StatusStarting: true,
+			host.StatusRunning:  true,
+		},
+	}
+}
+
+func main() {
+	return
+}
+
+func (s *Scheduler) Run() error {
+	log := s.log.New("fn", "Run")
+	log.Info("starting scheduler loop")
+	defer log.Info("exiting scheduler loop")
+
+	for {
+		// first, check if we should stop or process pending job events
+		select {
+		case <-s.stop:
+			return nil
+		case req := <-s.jobRequests:
+			s.HandleJobRequest(req)
+			continue
+		default:
+		}
+
+		log.Info("starting cluster sync")
+		if err := s.Sync(); err != nil {
+			log.Error("error performing cluster sync", "err", err)
+			continue
+		}
+
+		log.Info("starting watching events")
+		select {
+		case <-s.stop:
+			return nil
+		case fc := <-s.formationChange:
+			if err := s.FormationChange(fc); err != nil {
+				log.Error("error performing formation change", "err", err)
+				continue
+			}
+		case <-time.After(time.Second):
+		}
+	}
+	return nil
+}
+
+func (s *Scheduler) Sync() (err error) {
+	log := s.log.New("fn", "Sync")
+
+	defer func() {
+		s.sendEvent(NewEvent(EventTypeClusterSync, err, nil))
+	}()
+
+	log.Info("getting host list")
+	hosts, err := s.Hosts()
+	if err != nil {
+		log.Error("error getting host list", "err", err)
+		return err
+	}
+	log.Info(fmt.Sprintf("got %d hosts", len(hosts)))
+
+	for _, h := range hosts {
+		log = log.New("host_id", h.ID())
+		log.Info("getting jobs list")
+		activeJobs, err := h.ListJobs()
```

Comment:
`err` here is local to the for loop, so will not be returned when the `if err != nil` check is made outside the loop. You should explicitly declare `activeJobs` and use `activeJobs, err = h.ListJobs()`